### ISSUE TITLE
kubernetes deployment does not delete the old ReplicaSet

### ISSUE THREAD

**Issue by rata** [zendesk/samson#831]
Hi,

When testing the kubernetes plugin, using a 1.2.0 kubernetes cluster, I run into some issues.

The first deployment is done just fine. But when I do a new one, going also to "release" --> new release, etc. The deployment is made, but the old pods continue to be there running because the old replica set is there too. For example, after doing a second deploy, I see:
```
(env-aws) Rodrigos-MacBook-Pro:wido rodrigocampos$ kubectl get rs 
NAME                     DESIRED   CURRENT   AGE
wido-web-rc-3517251930   2         2         15m
wido-web-rc-3998679387   2         2         14m
(env-aws) Rodrigos-MacBook-Pro:wido rodrigocampos$ kubectl get deployments
NAME          DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
wido-web-rc   2         2         2            2           15m
(env-aws) Rodrigos-MacBook-Pro:wido rodrigocampos$ kubectl get pods
NAME                           READY     STATUS    RESTARTS   AGE
cadvisor-signalfx-qw9am        1/1       Running   1          2d
wido-web-rc-3517251930-0wg6t   3/3       Running   0          13m
wido-web-rc-3517251930-b7608   3/3       Running   0          15m
wido-web-rc-3998679387-sa9u2   3/3       Running   0          15m
wido-web-rc-3998679387-u5teb   3/3       Running   0          15m
```

The old pods are never deleted because the old ReplicaSet is never deleted. I'm not sure why this is, as the deployment is made server-side using the "Deployment" Kind.

Now I realize that maybe it is not detecting it is a new deploy and it is creating it again, and that's why it creates a new RS instead of having both at the same time and do a rolling update until the deploy finishes and the other ReplicaSet is deleted. When the deployment was running, the 2 new pods were created without waiting for the other to go down, so probably is not doing an update.

Is this possible?

**Comment by grosser** [comment]
@jonmoter

**Comment by jonmoter** [comment]
Hi Rata.  I'm looking into the same code right now.

The Samson/Kubernetes integration was done and tested on Kubernetes 1.1, and since the Deployments API is still in Beta, I think they might've changed a few things between 1.1 and 1.2.  I'm currently going through the code and figuring out what (if anything) needs to be updated.

There are also some issues with the Kubernetes Dashboard page when running against 1.2, which I'm looking at.  I'll cc you on any PRs I make to resolve it.

Thanks.

**Comment by rata** [comment]
@jonmoter great, thanks!

One other thing that needs updating is the yml file used for tests, the selector syntax changed. And there is no need to put a selector anyways (but plugins/kubernetes/app/models/concerns/kubernetes/deploy_yaml.rb needs to be updated in set_selector_metadata() as it assumes it always be there, but that is not true)

**Comment by rata** [comment]
@jonmoter let me know if I can help you!

**Comment by rata** [comment]
@jonmoter: Debugging more, the issue is caused because we commented the set_selector_metadata call in plugins/kubernetes/app/models/concerns/kubernetes/deploy_yaml.rb because the format changed in 1.2 (it needs to use matchselectors now, to work as it was working before, as in k8s 1.2 other ways to specify selectors were added).

Instead of commenting it, and fixing the set_selectors to use the spec.template.matchselectors.selector instead, fixes the issue. Will do a PR soon, maybe a friend will do it and reference this issue.


Thanks a lot!
